File: server/src/db/ledger.js
import { withTransaction } from './index.js';
import {
  dbTransactionDuration,
  dbTransactionErrorsTotal,
  handStartDuration,
  settlementDuration,
  safeLabel,
  timed,
} from '../metrics/index.js';

/**
 * Every chip movement a hand makes, each as one PostgreSQL transaction.
 *
 * This is the shape the game's money follows:
 *
 *   validate in memory (turn, amount, balance)
 *     → BEGIN
 *     → lock the wallet row            SELECT … FOR UPDATE
 *     → deduct the bet                 UPDATE users
 *     → add it to the pot              UPDATE pots
 *     → append the ledger entry        INSERT chip_ledger   (action_id UNIQUE)
 *     → save the table state/version   UPSERT game_states   (version must rise)
 *     → COMMIT
 *   → only then does the table change what it holds in memory and broadcast.
 *
 * Anything that fails inside the transaction rolls back completely, throws a
 * `LedgerError` with a code the table can turn into a `GameError`, and leaves
 * the game state untouched.
 */
export class LedgerError extends Error {
  constructor(code, message) {
    super(message);
    this.name = 'LedgerError';
    this.code = code;
  }
}

const now = () => Date.now();

/** PostgreSQL's unique_violation. */
const UNIQUE_VIOLATION = '23505';

/**
 * Turns a low-level failure into a LedgerError the caller can reason about.
 * The action_id UNIQUE index is what makes a retried move harmless, and a
 * violation on it is the one database error that is expected in normal play.
 */
const classify = (error) => {
  if (error instanceof LedgerError) return error;
  if (error?.code === UNIQUE_VIOLATION && /action_id/.test(error.detail ?? error.constraint ?? '')) {
    return new LedgerError('duplicate_action', 'That move has already been applied');
  }
  const wrapped = new LedgerError('persist_failed', error?.message ?? 'database write failed');
  wrapped.cause = error;
  return wrapped;
};

/** The codes a ledger operation can fail with; anything else is folded into "other". */
const KNOWN_LEDGER_CODES = new Set([
  'duplicate_action',
  'insufficient_chips',
  'stale_state',
  'no_pot',
  'unknown_user',
  'invalid_amount',
  'persist_failed',
]);

/**
 * Runs one ledger operation under its metrics: the transaction's duration by
 * operation, and — when it rolls back — a count by error code. Every failure
 * leaves here as a LedgerError, whatever it started out as.
 */
async function transact(op, fn) {
  try {
    return await timed(dbTransactionDuration, { op }, fn);
  } catch (error) {
    const refusal = classify(error);
    dbTransactionErrorsTotal.inc({ op, code: safeLabel(refusal.code, KNOWN_LEDGER_CODES) });
    throw refusal;
  }
}

/**
 * Locks a wallet row and returns its balance. Locking first, then reading, is
 * what stops two bets from the same account racing past the balance check.
 */
async function lockWallet(client, userId) {
  const { rows } = await client.query(
    'SELECT chips FROM users WHERE id = $1 FOR UPDATE',
    [userId],
  );
  if (rows.length === 0) throw new LedgerError('unknown_user', `unknown user ${userId}`);
  return rows[0].chips;
}

async function appendLedger(client, { userId, handId, actionId, delta, balance, reason, at }) {
  await client.query(
    `INSERT INTO chip_ledger (user_id, hand_id, action_id, delta, balance, reason, created_at)
     VALUES ($1, $2, $3, $4, $5, $6, $7)`,
    [userId, handId ?? null, actionId ?? null, delta, balance, reason, at],
  );
}

/**
 * Saves the table's snapshot, refusing to go backwards in version.
 *
 * A single process always presents a rising version, so the WHERE clause is a
 * no-op for it; its purpose is the other case — a second process that thinks
 * it owns the same room — whose stale write is rejected and rolled back.
 */
async function saveState(client, { roomId, handId, version, state, at }) {
  if (version === undefined || version === null) return;
  const { rowCount } = await client.query(
    `INSERT INTO game_states (room_id, hand_id, version, state, updated_at)
     VALUES ($1, $2, $3, $4::jsonb, $5)
     ON CONFLICT (room_id) DO UPDATE
        SET hand_id = EXCLUDED.hand_id,
            version = EXCLUDED.version,
            state = EXCLUDED.state,
            updated_at = EXCLUDED.updated_at
      WHERE game_states.version < EXCLUDED.version`,
    [roomId, handId ?? null, version, JSON.stringify(state ?? {}), at],
  );
  if (rowCount === 0) {
    throw new LedgerError('stale_state', `state version ${version} is not newer than the stored one`);
  }
}

/**
 * The transaction behind a chaal, raise or show: one player's chips into the
 * pot. Returns the wallet balance after the deduction, which the table adopts
 * as the truth.
 */
export async function bet({ userId, amount, roomId, handId, actionId, reason = 'bet', version, state }) {
  return transact('bet', async () => {
    if (!Number.isInteger(amount) || amount <= 0) {
      throw new LedgerError('invalid_amount', 'bet amount must be a positive integer');
    }

    return withTransaction(async (client) => {
      const at = now();

      const chips = await lockWallet(client, userId);
      if (chips < amount) {
        throw new LedgerError('insufficient_chips', `insufficient chips for ${userId}`);
      }
      const balance = chips - amount;

      await client.query('UPDATE users SET chips = $1, updated_at = $2 WHERE id = $3', [balance, at, userId]);

      const pot = await client.query(
        'UPDATE pots SET amount = amount + $1 WHERE hand_id = $2',
        [amount, handId],
      );
      if (pot.rowCount === 0) {
        throw new LedgerError('no_pot', `no open pot for hand ${handId}`);
      }

      await appendLedger(client, { userId, handId, actionId, delta: -amount, balance, reason, at });
      await saveState(client, { roomId, handId, version, state, at });

      // `persisted` tells the table how much of this stake the account has
      // already been debited — here, all of it — so settlement knows the only
      // movement left is the payout.
      return { balance, persisted: amount };
    });
  });
}

/**
 * Opens the pot and takes the boot from every participant, in one transaction.
 *
 * All wallet rows are locked in a fixed order (by user id) so two tables that
 * happen to share a player cannot deadlock against each other. If any one
 * player cannot cover the boot the whole start is refused and nothing is
 * charged — the table then sweeps them and tries again.
 */
export async function collectBoot({ roomId, handId, bootAmount, entries, version, state }) {
  // The boot transaction is what a hand start costs, so the same span feeds
  // both the ledger histogram and the hand-start one.
  return timed(handStartDuration, {}, () => transact('boot', () =>
    withTransaction(async (client) => {
      const at = now();
      const ordered = [...entries].sort((a, b) => (a.userId < b.userId ? -1 : 1));
      const balances = {};

      for (const { userId, amount } of ordered) {
        const chips = await lockWallet(client, userId);
        if (chips < amount) {
          const error = new LedgerError('insufficient_chips', `insufficient chips for ${userId}`);
          error.userId = userId;
          throw error;
        }
        balances[userId] = chips - amount;
        await client.query('UPDATE users SET chips = $1, updated_at = $2 WHERE id = $3', [balances[userId], at, userId]);
      }

      const total = ordered.reduce((sum, entry) => sum + entry.amount, 0);
      await client.query(
        `INSERT INTO pots (hand_id, room_id, boot_amount, amount, opened_at)
         VALUES ($1, $2, $3, $4, $5)`,
        [handId, roomId, bootAmount, total, at],
      );

      for (const { userId, amount } of ordered) {
        await appendLedger(client, {
          userId,
          handId,
          // Deterministic, so a retried start cannot ante the same player twice.
          actionId: `${handId}:boot:${userId}`,
          delta: -amount,
          balance: balances[userId],
          reason: 'boot',
          at,
        });
      }

      await saveState(client, { roomId, handId, version, state, at });
      return { balances, persisted: bootAmount };
    })));
}

/**
 * Settles a whole hand in one transaction: the hand record, the winner's
 * payout (every stake was already banked as it was bet, so the only movement
 * left is the pot going to the winner), the play/win counters, the pot
 * closing, and the final state snapshot.
 *
 * `entries` is `[{ userId, delta, isWinner, didChaal, leftMidHand }]` with
 * deltas already computed by the table. Returns `{ userId: balance }` for
 * every entry, which the table adopts.
 */
export async function settle({ hand, entries, version, state }) {
  return timed(settlementDuration, {}, () => transact('settle', () =>
    withTransaction(async (client) => {
      const at = now();

      await client.query(
        `INSERT INTO hands (id, room_id, hand_no, pot, winner_id, win_reason,
                            boot_amount, started_at, ended_at, summary_json)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10::jsonb)
         ON CONFLICT (id) DO NOTHING`,
        [
          hand.id,
          hand.roomId,
          hand.handNo,
          hand.pot,
          hand.winnerId ?? null,
          hand.winReason ?? null,
          hand.bootAmount,
          hand.startedAt,
          hand.endedAt,
          JSON.stringify(hand.summary ?? []),
        ],
      );

      const balances = {};
      const ordered = [...entries].sort((a, b) => (a.userId < b.userId ? -1 : 1));

      for (const entry of ordered) {
        const { rows } = await client.query('SELECT chips FROM users WHERE id = $1 FOR UPDATE', [entry.userId]);
        if (rows.length === 0) continue;

        // Never below zero: a delta here is a payout or a correction, and the
        // wallet CHECK would reject an overdraft anyway.
        const balance = Math.max(0, rows[0].chips + entry.delta);

        // "Played" means the player committed chips beyond the boot — posting
        // the ante and folding straight away is not a hand played.
        const played = entry.didChaal ? 1 : 0;
        const left = entry.leftMidHand ? 1 : 0;
        const lost = !entry.isWinner && !entry.leftMidHand ? 1 : 0;

        await client.query(
          `UPDATE users
              SET chips          = $1,
                  hands_played   = hands_played + $2,
                  hands_won      = hands_won + $3,
                  hands_lost     = hands_lost + $4,
                  hands_left_mid = hands_left_mid + $5,
                  total_winnings = total_winnings + $6,
                  biggest_pot    = GREATEST(biggest_pot, $7),
                  updated_at     = $8
            WHERE id = $9`,
          [
            balance,
            played,
            entry.isWinner ? 1 : 0,
            lost,
            left,
            entry.isWinner ? hand.pot : 0,
            entry.isWinner ? hand.pot : 0,
            at,
            entry.userId,
          ],
        );

        // A zero delta is still recorded: the row is what says this player was
        // in the hand and how it ended for them.
        await appendLedger(client, {
          userId: entry.userId,
          handId: hand.id,
          actionId: `${hand.id}:settle:${entry.userId}`,
          delta: entry.delta,
          balance,
          reason: entry.isWinner ? 'hand_win' : 'hand_loss',
          at,
        });

        balances[entry.userId] = balance;
      }

      await client.query(
        'UPDATE pots SET closed_at = $1, winner_id = $2 WHERE hand_id = $3',
        [at, hand.winnerId ?? null, hand.id],
      );

      await saveState(client, { roomId: hand.roomId, handId: null, version, state, at });
      return balances;
    })));
}

/**
 * The bundle a Table is built with. Kept as a plain object so a test can hand
 * the table an in-memory stand-in with the same three methods.
 */
export function createLedger() {
  return { bet, collectBoot, settle };
}

export default { bet, collectBoot, settle, createLedger, LedgerError };

File: go-server/internal/db/ledger.go
package db

import (
	"context"
	"errors"
	"sort"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"

	"github.com/surajk543/king-teenpatti/go-server/internal/game"
	"github.com/surajk543/king-teenpatti/go-server/internal/metrics"
)

// UniqueViolation is PostgreSQL SQLSTATE 23505. A violation whose constraint
// or detail mentions action_id is the one database error expected in normal
// play: a retried move, refused as duplicate_action.
const UniqueViolation = "23505"

// Ledger is the PostgreSQL game.Ledger (db/ledger.js). Every method is one
// transaction under WithTx, timed into game_db_transaction_duration_seconds{op}
// and counted on failure in game_db_transaction_errors_total{op,code}; boot
// also feeds game_hand_start_duration_seconds and settle
// game_settlement_duration_seconds (Node wrapped them the same way).
//
// Every error leaves here as a *game.GameError with a KnownLedgerCodes code
// (Classify); the driver error is kept in Cause for logs.
type Ledger struct {
	db      *DB
	metrics *metrics.Metrics // may be nil (tests)
	clock   func() time.Time
}

// NewLedger builds the ledger. m may be nil (no observations); clock nil →
// time.Now.
func NewLedger(d *DB, m *metrics.Metrics, clock func() time.Time) *Ledger {
	return &Ledger{db: d, metrics: m, clock: clock}
}

var _ game.Ledger = (*Ledger)(nil)

// Checkpoint implements game.Ledger: ONE player's pack or leave checkpoint,
// in one transaction.
//
//	SELECT chips FROM users WHERE id = $1 FOR UPDATE          (no row → unknown_user)
//	UPDATE users SET chips = chips + delta, counters…, updated_at
//	INSERT INTO chip_ledger (…, action_id UNIQUE)             (replay → duplicate_action, whole txn rolls back)
//
// The delta is what the Table computed (`chips now − chips as last written`);
// this never writes an absolute value, so a reward credited to a seated
// player is not erased. A zero delta still writes its row.
func (l *Ledger) Checkpoint(ctx context.Context, req game.CheckpointRequest) (game.CheckpointResult, error) {
	var result game.CheckpointResult
	err := l.transact(metrics.OpCheckpoint, func() error {
		return l.db.WithTx(ctx, func(tx pgx.Tx) error {
			balance, err := applyCheckpoint(ctx, tx, req.Entry, req.HandID, now(l.clock))
			if errors.Is(err, errAccountGone) {
				// Settle skips a deleted account silently (its rows went with
				// it); a single-player checkpoint has nothing else to do, so
				// it says so.
				return game.Errorf(game.CodeUnknownUser, "unknown user %s", req.Entry.UserID)
			}
			if err != nil {
				return err
			}
			result = game.CheckpointResult{Balance: balance}
			return nil
		})
	})
	if err != nil {
		return game.CheckpointResult{}, err
	}
	return result, nil
}

// applyCheckpoint writes one player's row against a wallet it locks itself,
// and returns the balance it left. An account that no longer exists is
// skipped silently (Node's `continue`), reporting a balance of 0.
func applyCheckpoint(ctx context.Context, tx pgx.Tx, entry game.SettleEntry, handID string, at int64) (int64, error) {
	chips, err := lockWallet(ctx, tx, entry.UserID)
	if err != nil {
		if game.CodeOf(err, "") == game.CodeUnknownUser {
			return 0, errAccountGone
		}
		return 0, err
	}

	// Never below zero: the CHECK on users.chips would reject an overdraft,
	// and a checkpoint is a correction, not a bet.
	balance := chips + entry.Delta
	if balance < 0 {
		balance = 0
	}

	// Counters land on the row that RESOLVES the hand for this player: the
	// settlement row, or the leave row for someone who walked out. A pack
	// checkpoint moves money only. "Played" means chips beyond the boot —
	// posting the ante and folding straight away is not a hand played
	// (requirement 16).
	played, won, lost, left := 0, 0, 0, 0
	var gross int64
	if entry.Outcome {
		played = boolToInt(entry.DidChaal)
		won = boolToInt(entry.IsWinner)
		lost = boolToInt(!entry.IsWinner && !entry.LeftMidHand)
		left = boolToInt(entry.LeftMidHand)
		if entry.IsWinner {
			gross = entry.Pot
		}
	}

	if _, err := tx.Exec(ctx, `UPDATE users
          SET chips          = $1,
              hands_played   = hands_played + $2,
              hands_won      = hands_won + $3,
              hands_lost     = hands_lost + $4,
              hands_left_mid = hands_left_mid + $5,
              total_winnings = total_winnings + $6,
              biggest_pot    = GREATEST(biggest_pot, $7),
              updated_at     = $8
        WHERE id = $9`,
		balance, played, won, lost, left, gross, gross, at, entry.UserID); err != nil {
		return 0, err
	}

	// A zero delta is still recorded: the row is what says this player was in
	// the hand and how it ended for them.
	if err := appendLedger(ctx, tx, entry.UserID, handID, entry.ActionID, entry.Delta, balance, entry.Reason, at); err != nil {
		return 0, err
	}
	return balance, nil
}

// errAccountGone marks a checkpoint whose wallet row has been deleted: the
// entry is skipped, not failed (the ledger rows went with the account).
var errAccountGone = errors.New("account gone")

// Settle implements game.Ledger: the HAND-END checkpoint. ONE transaction,
// every entry through applyCheckpoint in ascending userId order — wallets
// locked in order so two tables sharing a player cannot deadlock, and that is
// now the ONLY lock this transaction takes (the `hands` table is gone, so
// there is no foreign-key lock on the winner's row to order around any more;
// DECISIONS §2's rule about inserting `hands` after the wallets is moot).
//
// The retry path (Table.retrySettle) resends exactly this request; a replay
// hits the per-entry action ids' UNIQUE index, rolls the whole transaction
// back and returns duplicate_action, which the Table reads as the success it
// is. That uniqueness is the settle-retry safety mechanism: a commit whose
// acknowledgement is lost must never pay the winner the pot twice.
func (l *Ledger) Settle(ctx context.Context, req game.SettleRequest) (game.SettleResult, error) {
	var result game.SettleResult
	started := time.Now()
	err := l.transact(metrics.OpSettle, func() error {
		return l.db.WithTx(ctx, func(tx pgx.Tx) error {
			at := now(l.clock)
			balances := make(game.SettleResult, len(req.Entries))
			ordered := make([]game.SettleEntry, len(req.Entries))
			copy(ordered, req.Entries)
			sort.SliceStable(ordered, func(i, j int) bool { return ordered[i].UserID < ordered[j].UserID })

			for _, entry := range ordered {
				balance, err := applyCheckpoint(ctx, tx, entry, req.HandID, at)
				if errors.Is(err, errAccountGone) {
					continue
				}
				if err != nil {
					return err
				}
				balances[entry.UserID] = balance
			}
			result = balances
			return nil
		})
	})
	if l.metrics != nil && l.metrics.SettlementDuration != nil {
		l.metrics.SettlementDuration.Observe(time.Since(started).Seconds())
	}
	if err != nil {
		return nil, err
	}
	return result, nil
}

// transact runs one ledger operation under its metrics (ledger.js transact):
// the transaction's duration by operation, and — when it fails — a count by
// error code. Every failure leaves here as a *game.GameError, whatever it
// started out as. fn is the whole operation including any pre-BEGIN
// validation, exactly the span Node timed.
func (l *Ledger) transact(op string, fn func() error) error {
	started := time.Now()
	err := fn()
	if l.metrics != nil && l.metrics.DBTransactionDuration != nil {
		l.metrics.DBTransactionDuration.WithLabelValues(op).Observe(time.Since(started).Seconds())
	}
	if err == nil {
		return nil
	}
	refusal := Classify(err)
	if l.metrics != nil && l.metrics.DBTransactionErrors != nil {
		code := metrics.SafeLabel(refusal.Code, game.KnownLedgerCodes, metrics.OtherLabel)
		l.metrics.DBTransactionErrors.WithLabelValues(op, code).Inc()
	}
	return refusal
}

// Classify turns any error into the *game.GameError the Table expects
// (ledger.js classify): a *game.GameError passes through; a pgconn.PgError
// with Code UniqueViolation whose ConstraintName or Detail contains
// "action_id" → duplicate_action ("That move has already been applied");
// anything else → persist_failed with the message of err and Cause = err.
// nil → nil.
func Classify(err error) *game.GameError {
	if err == nil {
		return nil
	}
	var ge *game.GameError
	if errors.As(err, &ge) {
		return ge
	}
	var pgErr *pgconn.PgError
	if errors.As(err, &pgErr) && pgErr.Code == UniqueViolation && containsAny("action_id", pgErr.Detail, pgErr.ConstraintName) {
		return game.NewGameError(game.CodeDuplicateAction, "That move has already been applied")
	}
	message := err.Error()
	if message == "" {
		message = "database write failed"
	}
	return &game.GameError{Code: game.CodePersistFailed, Message: message, Cause: err}
}

// lockWallet: SELECT chips … FOR UPDATE; unknown_user when no row. Locking
// first, then reading, is what stops two bets from the same account racing
// past the balance check.
func lockWallet(ctx context.Context, tx pgx.Tx, userID string) (int64, error) {
	var chips int64
	err := tx.QueryRow(ctx, `SELECT chips FROM users WHERE id = $1 FOR UPDATE`, userID).Scan(&chips)
	if errors.Is(err, pgx.ErrNoRows) {
		return 0, game.Errorf(game.CodeUnknownUser, "unknown user %s", userID)
	}
	if err != nil {
		return 0, err
	}
	return chips, nil
}

// appendLedger inserts one chip_ledger row. handID/actionID "" → NULL.
func appendLedger(ctx context.Context, tx pgx.Tx, userID, handID, actionID string, delta, balance int64, reason string, at int64) error {
	_, err := tx.Exec(ctx, `INSERT INTO chip_ledger (user_id, hand_id, action_id, delta, balance, reason, created_at)
     VALUES ($1, $2, $3, $4, $5, $6, $7)`,
		userID, nullIfEmpty(handID), nullIfEmpty(actionID), delta, balance, reason, at)
	return err
}

// containsAny reports whether any of the haystacks contains needle.
func containsAny(needle string, haystacks ...string) bool {
	for _, h := range haystacks {
		if strings.Contains(h, needle) {
			return true
		}
	}
	return false
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}
